src/ctia/schemas/utils.clj

```diff
@@ -18,55 +19,169 @@
    identity
    s))
 
-
 ;; fns below are backported from schema-tools master
 ;; from https://github.com/metosin/schema-tools/blob/master/src/schema_tools/core.cljc#L266
 ;; TODO those can be removed once we use 0.10.6
 
-(defn- explicit-key [k] (if (s/specific-key? k) (s/explicit-schema-key k) k))
-
-(defn- explicit-key-set [ks]
-  (reduce (fn [s k] (conj s (explicit-key k))) #{} ks))
-
-(defn- maybe-anonymous [original current]
-  (if (= original current)
-    original
-    (vary-meta
-     current
-     (fn [meta]
-       (let [new-meta (clojure.core/dissoc meta :name :ns)]
-         (if (empty? new-meta)
-           nil
-           new-meta))))))
-
-(defn- transform-keys
-  [schema f ks]
-  (assert (or (not ks) (vector? ks)) "input should be nil or a vector of keys.")
-  (maybe-anonymous
-   schema
-   (let [ks? (explicit-key-set ks)]
-     (stu/map-keys
-      (fn [k]
-        (cond
-          (and ks (not (ks? (explicit-key k)))) k
-          (s/specific-key? k) (f (s/explicit-schema-key k))
-          :else k))
-      schema))))
-
+;; TODO delete me
 (defn optional-keys
   "Makes given map keys optional. Defaults to all keys."
-  ([m] (optional-keys m nil))
-  ([m ks] (transform-keys m s/optional-key ks)))
+  ([m] (st/optional-keys m))
+  ([m ks] (st/optional-keys m ks)))
 
+;; TODO delete me
 (defn optional-keys-schema
   "Walks a schema making all keys optional in Map Schemas."
   [schema]
-  (sw/prewalk
-   (fn [x]
-     (if (and (map? x) (not (record? x)))
-       (optional-keys x)
-       x))
-   schema))
+  (st/optional-keys-schema schema))
+
+(s/defschema SpecificKey
+  (s/pred s/specific-key?))
+
+(s/defn service-subgraph :- (s/pred map?)
```

Comment:
I wonder if it would be easy to test the value of your second argument to have a single function polymorphic in the second argument type.

If a selector `{SpecificKey #{SpecificKey}}` then use this implementation.
If a service schema => then use the implementation from `service-subgraph-from-schema`.

I am generally against this. But here, as the usage will be quite specific, it would make sense to me to use the same function name in both case `service-subgraph` and let it decide for me if it should be a `service-subgraph-from-schema` or `service-subgraph-from-service-subgraph-selector`.

Reply:
I merged them as it's just a matter of testing for set vs. map. This also solves the performance problem I had commented in `service-subgraph-from-schema` where the schema was traversed twice--now it's a single pass.